### PR TITLE
Remove SpecialType handling from KnownType

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs [hunk 1]
```diff
@@ -24,100 +24,76 @@
...
+        private readonly IList<string> namespaceParts;
+        private readonly string[] genericParameters;
 
-        internal abstract bool Matches(ITypeSymbol symbol);
+        public string TypeName { get; }
+        public string FullName { get; }
+        public bool IsArray { get; init; }
 
-        internal sealed class RegularKnownType : KnownType
+        internal KnownType(string fullName, params string[] genericParameters)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:36]: All these `internal` should be `public` - forward from https://github.com/SonarSource/sonar-dotnet/pull/5689#discussion_r887949530

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs [hunk 2]
```diff
@@ -24,100 +24,76 @@
...
-            }
+            namespaceParts = new ArraySegment<string>(parts, 0, parts.Length - 1);
+        }
 
-            internal override bool Matches(ITypeSymbol symbol) =>
-                IsMatch(symbol)
-                || IsMatch(symbol.OriginalDefinition);
+        internal bool Matches(ITypeSymbol symbol) =>
+            IsMatch(symbol)
+            || IsMatch(symbol.OriginalDefinition);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:49]: Can be single line - forward https://github.com/SonarSource/sonar-dotnet/pull/5689#discussion_r887949927

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor stuff

**Comment by Corniel** [comment]
While at it, can you add a (better) debug experience, by implementing `ToString()` or adding a `[DebuggerDisplay]`?

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs [hunk 3]
```diff
@@ -25,90 +25,71 @@
...
+        public string TypeName { get; }
+        public string FullName { get; }
+        public bool IsArray { get; init; }
 
-        internal sealed class RegularKnownType : KnownType
+        public KnownType(string fullName, params string[] genericParameters)
         {
-            private readonly IList<string> namespaceParts;
-            private readonly string[] genericParameters;
+            FullName = fullName;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:39]: _Whitespace nitpick:_ I don't like the empty lines here, they are sort of irregular and doesn't help much.

Can we reorder it to have `var parts` as a first line? Then we could have just a series of initialization. In the same order as the fields & properties are declared above.
```
            var parts = fullName.Split('.');
            namespaceParts = new ArraySegment<string>(parts, 0, parts.Length - 1);
            this.genericParameters = genericParameters;
            TypeName = parts[parts.Length - 1];
            FullName = fullName;
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Comment by pavel-mikula-sonarsource** [comment]
> While at it, can you add a (better) debug experience, by implementing `ToString()` or adding a `[DebuggerDisplay]`?

That's a good idea 👍

**Comment by costin-zaharia-sonarsource** [comment]
> While at it, can you add a (better) debug experience, by implementing `ToString()` or adding a `[DebuggerDisplay]`?

Implemented :)

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs [hunk 4]
```diff
@@ -20,95 +20,76 @@
...
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using Microsoft.CodeAnalysis;
 
 namespace SonarAnalyzer.Helpers
 {
-    internal abstract partial class KnownType
+    [DebuggerDisplay("{FullName}, Array: {IsArray}, Parameters: {string.Join(\",\", genericParameters)}", Name = "{TypeName}")]
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:29]: Isn't this too hard to read? It dumps the technical internal state, but itsn't  very close to the more dev-friendly form of `System.Action<T>[]`?

It's meant to be in debug-only, so I'd be fine with `ToString()` and building it on the fly.

Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:29]: I wanted to keep things simple. We also want to be language-independent on KnownType and generating C# syntax for VB might be confusing.

Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:29]: > I wanted to keep things simple. We also want to be language-independent on KnownType and generating C# syntax for VB might be confusing.

In this case I would say that that would not be an issue; the VB.NET way of displaying types is not nice (understatement), and despite the context, reading a known type the C# way is - imho - the preferred way. Just my 2cts.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:29]: We agreed to improve it in next PR. C# is fine here, it's debug-only purposes

**Pull request merged**